src/lib/stream.rs

```diff
@@ -205,39 +157,140 @@ impl<'a> CodedInputStream<'a> {
         Ok(())
     }
 
-    fn recompute_buffer_size_after_limit(&mut self) {
-        self.buffer_size += self.buffer_size_after_limit;
-        let buffer_end = self.total_bytes_retired + self.buffer_size;
-        if buffer_end > self.current_limit {
-            // limit is in current buffer
-            self.buffer_size_after_limit = buffer_end - self.current_limit;
-            self.buffer_size -= self.buffer_size_after_limit;
-        } else {
-            self.buffer_size_after_limit = 0;
+    pub fn read_raw_bytes_immediate(&mut self, buf: &mut[u8]) -> ProtobufResult<()> {
+        let count = buf.len();
+        if count == 0 {
+            return Ok(())
+        }
+        assert!(count <= 4096);
+        while self.buffer().remaining() < count {
+            try!(self.refill_buffer_really());
+        }
+        let r = self.buffer().read_slice(buf);
+        assert_eq!(r, count);
+        Ok(())
+    }
+
+    pub fn read_raw_byte(&mut self) -> ProtobufResult<u8> {
+        if !self.buffer().has_remaining() {
+            try!(self.refill_buffer_really());
+        }
+        let buf = self.buffer();
+        let v = {
+            let view = Buf::bytes(buf);
+            unsafe { *view.get_unchecked(0) }
+        };
+        buf.advance(1);
+        Ok(v)
+    }
+
+    /// Read raw bytes into supplied slice, potentially many buffers long.
+    pub fn read_raw_bytes_into(&mut self, count: u32, target: &mut[u8]) -> ProtobufResult<()> {
+        if count == 0 {
+            return Ok(())
+        }
+        assert!(count as usize <= target.len());
```

Comment:
`count` parameter is redundant. `target` slice can be easily made of `count` size by caller.